core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,577 @@
+	pub fn get_mut(
+		&mut self,
+		states: &States,
+	) -> Option<HistoricalValue<&mut V>> {
+		let mut index = self.len();
+		if index == 0 {
+			return None;
+		}
+		debug_assert!(states.history.len() >= index);
+		let mut result = None;
+		let mut previous_transaction = usize::max_value();
+		let mut previous_switch = None;
+		while index > 0 {
+			index -= 1;
+			let state_index = self.get_state(index).index;
+			match states.history[state_index] {
+				TransactionState::TxPending => {
+					if state_index >= previous_transaction {
+						previous_switch = Some((index, state_index));
+					} else {
+						if result.is_none() {
+							result = Some((index, state_index));
+						}
+					}
+					break;
+				},
+				TransactionState::Pending => {
+					if state_index >= previous_transaction {
+						previous_switch = Some((index, state_index));
+					} else {
+						if result.is_none() {
+							result = Some((index, state_index));
+							previous_transaction = find_previous_tx_start(states, state_index);
+						} else {
+							break;
+						}
+					}
+				},
+				TransactionState::Dropped => (),
+			}
+		}
+		self.clear_terminal_values(result, previous_switch, 0)
+	}
+
+
+	/// Method to prune regarding a full state.
+	/// It also returns the last value as mutable.
+	/// Internally it acts like `get_mut` with an
+	/// additional cleaning capacity to clean committed
+	/// state if `prune_to_commit` is set to true.
+	pub fn get_mut_pruning(
+		&mut self,
+		states: &States,
+		prune_to_commit: bool,
+	) -> Option<HistoricalValue<&mut V>>  {
+		let mut index = self.len();
+		if index == 0 {
+			return None;
+		}
+		let mut prune_index = 0;
+		debug_assert!(states.history.len() >= index);
+		let mut result = None;
+		let mut previous_transaction = usize::max_value();
+		let mut previous_switch = None;
+		while index > 0 {
+			index -= 1;
+			let state_index = self.get_state(index).index;
+			match states.history[state_index] {
+				state @ TransactionState::TxPending
+				| state @ TransactionState::Pending => {
+					if state_index < states.commit && index > prune_index {
+						prune_index = index;
+					}
+
+					if state_index >= previous_transaction {
+						previous_switch = Some((index, state_index));
+					} else {
+						if result.is_none() {
+							result = Some((index, state_index));
+							if state == TransactionState::Pending {
+								previous_transaction = find_previous_tx_start(states, state_index);
+							}
+						} else {
+							if prune_to_commit {
+								if state_index < states.commit {
+									break;
+								}
+							} else {
+								break;
+							}
+						}
+					}
+				},
+				TransactionState::Dropped => (),
+			}
+		}
+		let deleted = if prune_to_commit && prune_index > 0 && result.is_some() {
+			self.truncate_until(prune_index);
+			prune_index
+		} else {
+			0
+		};
+		self.clear_terminal_values(result, previous_switch, deleted)
+	}
+
+	#[inline]
```

Comment:
That is probably not useful. I did add this function to avoid code redundancy but does not want a new function at all, so from bad habit I added it.
Last parameter is constant in one of its two calls, but lets trust the compiler.